tests/subscriptions/test_worker.py

```diff
@@ -0,0 +1,127 @@
+    def __init__(self, tasks: Iterable[ScheduledTask[Subscription]]):
+        self.__tasks = tasks
+
+    def find(
+        self, interval: Interval[datetime]
+    ) -> Iterator[ScheduledTask[Subscription]]:
+        for task in self.__tasks:
+            yield task
+
+
+class TestSubscriptionWorker(BaseSubscriptionTest):
+    def test(self):
+        result_topic = Topic("subscription-results")
+
+        broker: DummyBroker[SubscriptionTaskResult] = DummyBroker()
+        broker.create_topic(result_topic, partitions=1)
+
+        subscription = Subscription(
+            SubscriptionIdentifier(PartitionId(0), uuid1()),
+            SubscriptionData(
+                project_id=self.project_id,
+                conditions=[["platform", "IN", ["a"]]],
+                aggregations=[["count()", "", "count"]],
+                time_window=timedelta(minutes=500),
+                resolution=timedelta(minutes=1),
+            ),
+        )
+
+        now = datetime.utcnow()
+
+        timestamps = [
+            now,
+            (now + timedelta(minutes=self.minutes) + subscription.data.time_window),
+        ]
+
+        scheduler = DummyScheduler(
+            ScheduledTask(timestamp, subscription) for timestamp in timestamps
+        )
+
+        worker = SubscriptionWorker(
+            self.dataset,
+            ThreadPoolExecutor(),
+            {0: scheduler},
+            DummyProducer(broker),
+            result_topic,
+            DummyMetricsBackend(strict=True),
+        )
+
+        # Process the tick message and run the scheduled queries.
+        results = worker.process_message(
+            Message(
+                Partition(Topic("events"), 0),
+                0,
+                Tick(  # XXX: isn't actually used in the dummy scheduler
+                    offsets=Interval(0, 1),
+                    timestamps=Interval(timestamps[0], timestamps[-1]),
+                ),
+                timestamps[-1],
+            ),
+        )
+
+        assert results is not None
+
+        # Publish the results.
+        worker.flush_batch([results])
+
+        # Check to make sure the results were published.
+        # NOTE: This does not cover the ``SubscriptionTaskResultCodec``!
+        consumer: Consumer[SubscriptionTaskResult] = DummyConsumer(broker, "group")
+        consumer.subscribe([result_topic])
+
+        message = consumer.poll()
+        assert message is not None
+        assert message.partition.topic == result_topic
+        assert message.payload.task == results[0].task
+        assert message.payload.task.timestamp == timestamps[0]
+        assert message.payload.result == results[0].future.result()
+        request, result = message.payload.result
+        assert request.extensions["timeseries"] == {
+            "from_date": (timestamps[0] - subscription.data.time_window).isoformat(),
+            "to_date": timestamps[0].isoformat(),
+            "granularity": 3600,  # XXX: unused, no time grouping
+        }
```

Comment:
It does seem like these responsibilities should be split (constraint enforcement and bucketing.)